syntax = "proto3";

package rollercoaster;

option go_package = "rollercoaster/proto";

import "google/protobuf/empty.proto";

message StatusResponse {
	int32 total_wagons = 1;
	int32 total_passengers = 2;
	int32 waiting_passengers = 3;
}

service wagon {
	// Wagon stationed and waiting to be filled (blocks until ready)
	rpc stationed(google.protobuf.Empty) returns (google.protobuf.Empty);
	rpc depart(passenger_list) returns (google.protobuf.Empty);
	// finished ride -> start disembarking
	rpc arrive(google.protobuf.Empty) returns (google.protobuf.Empty);
}

service passenger {
	rpc i_am_disembarking(google.protobuf.Empty) returns (google.protobuf.Empty);
	rpc i_am_boarding(google.protobuf.Empty) returns (google.protobuf.Empty);
}

service rollercoaster {
	rpc get_status(google.protobuf.Empty) returns (StatusResponse);
}

message passenger_list {
	repeated int32 passenger_id = 1;
}
